cmd/notation/plugin/install.go

```diff
@@ -0,0 +1,329 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/notaryproject/notation-go/dir"
+	"github.com/notaryproject/notation-go/log"
+	notationplugin "github.com/notaryproject/notation/cmd/notation/internal/plugin"
+	"github.com/notaryproject/notation/internal/cmd"
+	"github.com/notaryproject/notation/internal/osutil"
+	"github.com/spf13/cobra"
+)
+
+const (
+	notationPluginTmpDir = "notationPluginTmpDir"
+)
+
+type pluginInstallOpts struct {
+	cmd.LoggingFlagOpts
+	pluginSourceType notationplugin.PluginSourceType
+	pluginSource     string
+	inputCheckSum    string
+	isFile           bool
+	isUrl            bool
+	force            bool
+}
+
+func pluginInstallCommand(opts *pluginInstallOpts) *cobra.Command {
+	if opts == nil {
+		opts = &pluginInstallOpts{}
+	}
+	command := &cobra.Command{
+		Use:     "install [flags] <--file|--url> <plugin_source>",
+		Aliases: []string{"add"},
+		Short:   "Install plugin",
+		Long: `Install a plugin
+
+Example - Install plugin from file system:
+  notation plugin install --file wabbit-plugin-v1.0.zip
+
+Example - Install plugin from file system with user input SHA256 checksum:
+  notation plugin install --file wabbit-plugin-v1.0.zip --checksum abcdef 
+
+Example - Install plugin from file system regardless if it's already installed:
+  notation plugin install --file wabbit-plugin-v1.0.zip --force
+
+Example - Install plugin from file system with .tar.gz:
+  notation plugin install --file wabbit-plugin-v1.0.tar.gz
+
+Example - Install plugin from URL, SHA256 checksum is required:
+  notation plugin install --url https://wabbit-networks.com/intaller/linux/amd64/wabbit-plugin-v1.0.tar.gz --checksum abcxyz
+`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return errors.New("missing plugin file path or URL")
+			}
+			opts.pluginSource = args[0]
+			return nil
+		},
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			if opts.isFile {
+				opts.pluginSourceType = notationplugin.PluginSourceTypeFile
+				return nil
+			}
+			if opts.isUrl {
+				opts.pluginSourceType = notationplugin.PluginSourceTypeURL
+				return nil
+			}
+			return errors.New("must choose one and only one flag from [--file, --url]")
```

Comment:
nit: You may want to reword it for a better prompt.

Reply:
> Can we use [`MarkFlagsMutuallyExclusive`](https://github.com/spf13/cobra/issues/1216) ?

@priteshbandi yes, I have done that. The prompt here is for none of the above flags is set.

Reply:
> nit: You may want to reword it for a better prompt.

In Cobra's new version (v1.8.0), there's the new [MarkFlagsOneRequired](https://pkg.go.dev/github.com/spf13/cobra@v1.8.0#Command.MarkFlagsOneRequired). Updated to use it along with `MarkFlagsMutuallyExclusive`.